src/boot/efi/boot.c

```diff
@@ -1871,6 +1872,115 @@ static bool is_sd_boot(EFI_FILE *root_dir, const char16_t *loader_path) {
+                .fv = {
+                        .Header = {
+                                .Type = MEDIA_FILEPATH_DP,
+                                .SubType = MEDIA_PIWG_FW_VOL_DP,
+                                .Length = sizeof(internal_uefi_shell_dp.fv),
+                        },
+                },
+                .ff = {
+                        .Header = {
+                                .Type = MEDIA_FILEPATH_DP,
+                                .SubType = MEDIA_PIWG_FW_FILE_DP,
+                                .Length = sizeof(internal_uefi_shell_dp.ff),
+                        },
+                        .Guid = UEFI_SHELL_GUID,
+                },
+                .end = DEVICE_PATH_END_NODE,
+        };
+
+        internal_uefi_shell_dp.fv.Guid = dp->Guid;
+
+        _cleanup_(unload_imagep) EFI_HANDLE shell;
+        err = BS->LoadImage(false, IMG, (EFI_DEVICE_PATH *) &internal_uefi_shell_dp, NULL, 0, &shell);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Error loading internal UEFI shell: %m");
+
+        err = BS->StartImage(shell, NULL, NULL);
+        if (err != EFI_SUCCESS)
+                return log_error_status(err, "Error starting internal UEFI shell: %m");
+
+        return EFI_SUCCESS;
+}
+
+static void config_entry_add_loader_uefi_internal(Config *config) {
+        EFI_STATUS err;
+
+        /* Finds the internal UEFI shell of EDK2 firmware images by locating it on the firmware volume.
+         * As the firmware volume GUID differs by architecture we simply iterate over all volumes
+         * and check if it contains the UEFI shell file.
+         *
+         * On x86 the full device path for the internal UEFI shell is:
+         * Fv(7CB8BDC9-F8EB-4F34-AAEA-3EE4AF6516A1)/FvFile(7C04A583-9E3E-4F1C-AD65-E05268D0B4D1)
+         *
+         * Once located, it is merely a matter to create the device path and pass it to LoadImage().
+         */
+
+        assert(config);
+
+        if (!config->auto_entries || !config->auto_firmware)
+                return;
+
+        EFI_HANDLE *handles;
+        size_t n_handles;
+        err = BS->LocateHandleBuffer(
+                        ByProtocol, MAKE_GUID_PTR(EFI_FIRMWARE_VOLUME2_PROTOCOL), NULL, &n_handles, &handles);
+        if (err != EFI_SUCCESS)
+                return;
+
+        for (size_t i = 0; i < n_handles; i++) {
+                EFI_FIRMWARE_VOLUME2_PROTOCOL *fv;
+                err = BS->HandleProtocol(
+                                handles[i], MAKE_GUID_PTR(EFI_FIRMWARE_VOLUME2_PROTOCOL), (void **) &fv);
+                if (err != EFI_SUCCESS)
+                        continue;
+
+                err = fv->ReadFile(
+                                fv,
+                                MAKE_GUID_PTR(UEFI_SHELL),
+                                NULL,
+                                &(size_t){ 0 },
+                                &(uint8_t){ 0x09 }, // EFI_FV_FILETYPE_APPLICATION
```

Comment:
as per coding style we prefer `/* */` for explanatory comments, and // for fixme style coments